hail/python/test/hailtop/batch/test_batch.py

```diff
@@ -1319,3 +1315,80 @@ def test_wait_on_empty_batch_update(self):
+        b = self.batch()
+        j = b.new_job()
+        for input in ["hi.txt", "~/hello.csv", "./hey.tsv", "/sup.json", "file://yo.yaml"]:
+            with pytest.raises(ValueError) as e:
+                b.read_input(input)
+            assert str(e.value).startswith("Local filepath detected")
+
+    @skip_in_azure
+    def test_validate_cloud_storage_policy(self):
+        # constants
+        gcs, fake_bucket1, fake_bucket2, no_bucket_error, no_perms_bucket, no_perms_error, cold_bucket, cold_error = [
+            GoogleStorageAsyncFS.cloud_name,
+            # buckets do not exist (bucket names can't contain the string "google" per
+            # https://cloud.google.com/storage/docs/buckets)
+            "google",
+            "google1",
+            "bucket does not exist",
+            # bucket exists, but account does not have permissions on it
+            "test",
+            "does not have storage.buckets.get access",
+            # bucket exists and account has permissions, but is set to use cold storage by default
+            "hail-test-cold-storage",
+            "Error while accessing",
+        ]
+        fake_uri1, fake_uri2, no_perms_uri, cold_uri = [
+            f"gs://{bucket}/test" for bucket in [fake_bucket1, fake_bucket2, no_perms_bucket, cold_bucket]
+        ]
+
+        # no configuration, nonexistent buckets error
+        with pytest.raises(ClientResponseError) as e:
+            ServiceBackend(remote_tmpdir=fake_uri1)
+        assert no_bucket_error in str(e.value)
+        with pytest.raises(ClientResponseError) as e:
+            ServiceBackend(remote_tmpdir=fake_uri2)
+        assert no_bucket_error in str(e.value)
+
+        # no configuration, no perms bucket errors
+        with pytest.raises(ClientResponseError) as e:
+            ServiceBackend(remote_tmpdir=no_perms_uri)
+        assert no_perms_error in str(e.value)
+
+        # no configuration, cold bucket errors
+        with pytest.raises(ValueError) as e:
+            ServiceBackend(remote_tmpdir=cold_uri)
+        assert str(e.value).startswith(cold_error)
+        b = self.batch()
+        with pytest.raises(ValueError) as e:
+            b.read_input(cold_uri)
+        assert str(e.value).startswith(cold_error)
+        j = b.new_job()
+        j.command(f"echo hello > {j.ofile}")
+        with pytest.raises(ValueError) as e:
+            b.write_output(j.ofile, cold_uri)
+        assert str(e.value).startswith(cold_error)
+
+        # hailctl config, allowlisted nonexistent buckets don't error
+        user_config.user_config.read_dict(
+            {"storage": {"location_allow_list": f"{gcs}:{fake_bucket1},{gcs}:{fake_bucket2}"}}
+        )
+        ServiceBackend(remote_tmpdir=fake_uri1)
+        ServiceBackend(remote_tmpdir=fake_uri2)
+
+        # environment variable config, only allowlisted nonexistent buckets don't error
+        os.environ["HAIL_STORAGE_LOCATION_ALLOW_LIST"] = f"{gcs}:{fake_bucket2}"
```

Comment:
I'm not sure we can safely test this environment variable here. Different tests running concurrently still run in the same process AFAIK so will share the same environment. So this could cause some spooky action if we start depending on this in multiple tests.

Reply:
Yeah, this is gonna be a pain to maintain. My suggestion:

1. Use explicit arguments for testing the various classes
2. Factor the parsing code into a tiny function and unit test that (if you want to test the parsing).

And we'll just accept the risk that `configuration_of` doesn't quite work right.

Reply:
okay, so i rewrote this to use [pytest's builtin monkeypatch fixture](https://docs.pytest.org/en/stable/how-to/monkeypatch.html#monkeypatching-environment-variables), which mocks the env for just this test; thoughts?